### PR TITLE
Entity generalization

### PR DESCRIPTION
**Pull request by VISHNUDAS-tunerlabs**
not for merge. For code review

### CODE REVIEW

File: generics/constants/common.js [hunk 1]
```diff
@@ -66,5 +66,8 @@ module.exports = {
   "LINK" : "link",
   "DISCOVERED_BY_ME" : "discoveredByMe",
   "DEFAULT_SUBMISSION_REQUIRED" : 1,
-  "SCHOOL" : "school"
+  "SCHOOL" : "school",
+  "SUBENTITY" :"subEntity_",
+  "SUNBIRD_SERVER_TIMEOUT" : 5000,
+  "TIMEOUT_ERROR" : "Server timeout error"
```
Comment (priyanka-TL) [inline generics/constants/common.js:72]: @VISHNUDAS-tunerlabs if its a response status/message add it in generics/https-status-codes

File: generics/services/form.js [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+  * @param {object} entityKey -  key to store data in cache
+  * @returns {Promise} returns a promise.
+*/
+
+
+const formData = function ( stateLocationCode, entityKey ) {
+    return new Promise(async (resolve, reject) => {
+        try {
+            
+            let subEntitiesDetails = await sunbirdService.formRead( stateLocationCode );
```
Comment (priyanka-TL) [inline generics/services/form.js:26]: @VISHNUDAS-tunerlabs handle the error also

File: generics/services/form.js [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+            let subEntities = subEntityData.map( subEntity => {
+                return subEntity.code;
+            })
+            if( !subEntitiesDetails.data || !subEntityData.length > 0 ) {
+                return resolve({
+                    message : constants.apiResponses.ENTITY_NOT_FOUND,
+                    result : []
+                })
+            }
+            let setCache = cache.setValue(entityKey, subEntities, 43200);
```
Comment (priyanka-TL) [inline generics/services/form.js:38]: @VISHNUDAS-tunerlabs set the cache time period in common file

File: generics/services/sunbird.js [hunk 4]
```diff
@@ -0,0 +1,211 @@
...
+*/
+
+const learnerLocationSearch = function ( filterData, limit = "", offset = ""  ) {
+  return new Promise(async (resolve, reject) => {
+      try {
+
+        let bodyData={};
+        bodyData["request"] = {};
+        bodyData["request"]["filters"] = filterData;
+        if( limit !== "" ) {
```
Comment (priyanka-TL) [inline generics/services/sunbird.js:31]: @VISHNUDAS-tunerlabs is limit and offset is same ?

File: generics/services/sunbird.js [hunk 5]
```diff
@@ -0,0 +1,211 @@
...
+            bodyData["request"]["limit"] = limit;
+        }
+        if( limit !== "" ) {
```
Comment (priyanka-TL) [inline generics/services/sunbird.js:34]: @VISHNUDAS-tunerlabs check offset is not empty

File: generics/services/sunbird.js [hunk 6]
```diff
@@ -0,0 +1,211 @@
...
+                    result["data"] = response.result;
+                } else {
+                      result.success = false;
+                }
+            }
+            return resolve(result);
+        }
+
+        setTimeout(function () {
+           return reject (constants.common.TIMEOUT_ERROR)
```
Comment (priyanka-TL) [inline generics/services/sunbird.js:73]: @VISHNUDAS-tunerlabs check value of success if its not passing make it false

File: generics/services/sunbird.js [hunk 7]
```diff
@@ -0,0 +1,211 @@
...
+        }, constants.common.SUNBIRD_SERVER_TIMEOUT);
```
Comment (priyanka-TL) [inline generics/services/sunbird.js:74]: @VISHNUDAS-tunerlabs if its a api response message add it in generics/api-responses

File: generics/services/sunbird.js [hunk 8]
```diff
@@ -0,0 +1,211 @@
...
+  * @param {object} bodyData -  subType data.
+  * @returns {Promise} returns a promise.
+*/
+const formRead = function ( subTypeData ) {
+    return new Promise(async (resolve, reject) => {
+        try {
+            
+            let bodyData = {
+                request : {
+                    type: "profileConfig",
```
Comment (priyanka-TL) [inline generics/services/sunbird.js:97]: @VISHNUDAS-tunerlabs add this in common

File: generics/services/sunbird.js [hunk 9]
```diff
@@ -0,0 +1,211 @@
...
+                    if( response.responseCode === constants.common.OK) {
+                        result["data"] = response.result;
+                    } else {
+                        result.success = false;
+                    }
+                }
+                return resolve(result);
+            }
+            setTimeout(function () {
+                return reject (constants.common.TIMEOUT_ERROR)
```
Comment (priyanka-TL) [inline generics/services/sunbird.js:136]: @VISHNUDAS-tunerlabs check value of success if its not passing make it false
 if its a api response message add it in generics/api-responses

File: generics/services/sunbird.js [hunk 10]
```diff
@@ -0,0 +1,211 @@
...
```
Comment (priyanka-TL) [inline generics/services/sunbird.js:198]: @VISHNUDAS-tunerlabs check value of success if its not passing make it false
 if its a api response message add it in generics/api-responses

File: module/programs/helper.js [hunk 11]
```diff
@@ -211,7 +212,7 @@ module.exports = class ProgramsHelper {
             {
               name : scopeData.entityType
             },
-            ["name","_id"]
+            ["name"]
```
Comment (priyanka-TL) [inline module/programs/helper.js:215]: @VISHNUDAS-tunerlabs no need to call entityType doc

File: module/programs/helper.js [hunk 12]
```diff
@@ -222,19 +223,29 @@ module.exports = class ProgramsHelper {
...
+            if (gen.utils.checkValidUUID(entity)) {
+              locationIds.push(entity);
+            }
+          });
+          bodyData = {
+            "id" : locationIds
+          }
+
+          
+          let entityData = await sunbirdService.learnerLocationSearch( bodyData );
```
Comment (priyanka-TL) [inline module/programs/helper.js:245]: @VISHNUDAS-tunerlabs handle the error

File: module/programs/helper.js [hunk 13]
```diff
@@ -721,37 +718,49 @@ module.exports = class ProgramsHelper {
...
+          }
+        });
+        bodyData = {
+          "id" : locationIds
+        }
+        
+        let entitiesData = await sunbirdService.learnerLocationSearch( bodyData );
+        
+
+        if( !entitiesData.data.count > 0 ) {
```
Comment (priyanka-TL) [inline module/programs/helper.js:745]: @VISHNUDAS-tunerlabs if api trigger timeout error "entitiesData.data" will present ?

File: module/solutions/helper.js [hunk 14]
```diff
@@ -119,57 +122,35 @@ module.exports = class SolutionsHelper {
...
-            let entitiesData = 
-            await entitiesHelper.entityDocuments({
-              _id : { $in : solutionData.entities }
-            },["_id"]);
+            let bodyData={
+              "id" : solutionData.entities
+            };
+            
 
+            let entitiesDetails = await sunbirdService.learnerLocationSearch( bodyData );
```
Comment (priyanka-TL) [inline module/solutions/helper.js:141]: @VISHNUDAS-tunerlabs handle the error also

File: module/solutions/helper.js [hunk 15]
```diff
@@ -261,65 +241,72 @@ module.exports = class SolutionsHelper {
...
+              scopeData.entities.forEach(entity => {
+                if (gen.utils.checkValidUUID(entity)) {
+                  entityLocationIds.push(entity);
+                }
+              });
+              let bodyData = {
+                "id" : entityLocationIds
+              };
+              
+              let entityDetails = await sunbirdService.learnerLocationSearch(bodyData);
```
Comment (priyanka-TL) [inline module/solutions/helper.js:265]: @VISHNUDAS-tunerlabs handle timeout error also

File: module/solutions/helper.js [hunk 16]
```diff
@@ -261,65 +241,72 @@ module.exports = class SolutionsHelper {
...
 
-            } else {
-              entityIds = entities.map(entity => {
-                return ObjectId(entity._id);
-              })
+              let entityIds = [];
+        
+              if( currentSolutionScope.entityType !== programData[0].scope.entityType ) {
+                let matchData = [];
+                let childList = await entitiesInParent(currentSolutionScope.entities,currentSolutionScope.entityType,matchData);
```
Comment (priyanka-TL) [inline module/solutions/helper.js:293]: @VISHNUDAS-tunerlabs change childList to childEntities

File: module/solutions/helper.js [hunk 17]
```diff
@@ -1022,40 +977,62 @@ module.exports = class SolutionsHelper {
...
         if( solutionData[0].scope.entityType !== programData[0].scope.entityType ) {
 
-          let checkEntityInParent = 
-          await entitiesHelper.entityDocuments({
-            _id : programData[0].scope.entities,
-            [`groups.${solutionData[0].entityType}`] : entities
-          },["_id"]);
+          let matchData = [];
+          let checkEntityInParent = [];
+          let childList = await entitiesInParent(programData[0].scope.entities,solutionData[0].scope.entityType,matchData);
```
Comment (priyanka-TL) [inline module/solutions/helper.js:984]: @VISHNUDAS-tunerlabs change childList to childEntities

File: module/solutions/helper.js [hunk 18]
```diff
@@ -2010,3 +1983,50 @@ function _generateLink(appsPortalBaseUrl, prefix, solutionLink, solutionType) {
...
+async function entitiesInParent( solutionEntities,currentEntityType,matchingData ){
+  
+  if( !solutionEntities.length > 0 ){
+    return matchingData;
+  }
+  let bodyData={
+    "parentId" : solutionEntities
+  };
+  
+  let entityDetails = await sunbirdService.learnerLocationSearch(bodyData);
```
Comment (priyanka-TL) [inline module/solutions/helper.js:2003]: @VISHNUDAS-tunerlabs handle if the data is empty check the success value first

File: module/users/helper.js [hunk 19]
```diff
@@ -152,15 +156,24 @@ module.exports = class UsersHelper {
...
-            },
-            ["entityType", "entityTypeId"]
-          );
+          let locationIds = [];
+          let bodyData={};
+          
 
-          if ( !entityData.length > 0 ) {
+          data.entities.forEach(entity=>{
+            if (gen.utils.checkValidUUID(entity)) {
```
Comment (priyanka-TL) [inline module/users/helper.js:166]: @VISHNUDAS-tunerlabs if its not a uuid then consider as school and check in org

File: module/users/helper.js [hunk 20]
```diff
@@ -152,15 +156,24 @@ module.exports = class UsersHelper {
...
+              locationIds.push(entity);
+            }
+          });
+        
+          bodyData = {
+            "id" : locationIds
+          } 
+          let entityData = await sunbirdService.learnerLocationSearch( bodyData );
```
Comment (priyanka-TL) [inline module/users/helper.js:174]: @VISHNUDAS-tunerlabs handle if the data is empty check the success value first

File: module/users/helper.js [hunk 21]
```diff
@@ -570,58 +584,62 @@ module.exports = class UsersHelper {
...
         if (!rolesDocument.length > 0) {
           throw {
             message: constants.apiResponses.USER_ROLES_NOT_FOUND
           };
         }
+        
 
+
+        let bodyData={};    
+        if (gen.utils.checkValidUUID(stateLocationId)) {
```
Comment (priyanka-TL) [inline module/users/helper.js:607]: @VISHNUDAS-tunerlabs this condition needs to be change according to type school and call orgs service

File: module/users/helper.js [hunk 22]
```diff
@@ -570,58 +584,62 @@ module.exports = class UsersHelper {
...
+          bodyData = {
+            "id" : stateLocationId
+          };
+        } else {
+          bodyData = {
+            "code" : stateLocationId
+          };
+        }
+        
+        let entityData = await sunbirdService.learnerLocationSearch( bodyData );
```
Comment (priyanka-TL) [inline module/users/helper.js:617]: @VISHNUDAS-tunerlabs handle if entityData is empty or handle timeout

File: module/users/helper.js [hunk 23]
```diff
@@ -686,82 +704,60 @@ module.exports = class UsersHelper {
...
+              "parentId" : requestedData[targetedEntityType]
+            }
           }
-
-          if (
-            entities[0] &&
-            entities[0].groups &&
-            Object.keys(entities[0].groups).length > 0
-          ) {
+          let entitiesData = await sunbirdService.learnerLocationSearch( filterData );
```
Comment (priyanka-TL) [inline module/users/helper.js:728]: @VISHNUDAS-tunerlabs handle if entityData is empty , check with success value

File: module/users/helper.js [hunk 24]
```diff
@@ -686,82 +704,60 @@ module.exports = class UsersHelper {
...
           };
         }
-
-        let entities = await entitiesHelper.entityDocuments(filterData, [
-          "metaInformation.name",
-          "entityType"
-        ]);
-
-        if (!entities.length > 0) {
+        let entitiesDocument = await sunbirdService.learnerLocationSearch( filterData );
```
Comment (priyanka-TL) [inline module/users/helper.js:743]: @VISHNUDAS-tunerlabs handle if entityData is empty , check with success value

File: module/users/helper.js [hunk 25]
```diff
@@ -782,60 +778,52 @@ module.exports = class UsersHelper {
...
-              _id: ObjectId(currentEntity._id)
-          }, ["groups"]);
-
-          if ( !entitiesDocument || !entitiesDocument.length > 0 ) {
-            continue;
-          }
-
-          entitiesDocument = entitiesDocument[0];
-          for( let entityCounter = 0 ; entityCounter < otherEntities.length ; entityCounter++ ) {
+            let entitiesData = await sunbirdService.learnerLocationSearch( filterData );
```
Comment (priyanka-TL) [inline module/users/helper.js:791]: @VISHNUDAS-tunerlabs handle if entityData is empty , check with success value

File: module/users/helper.js (hunk 21 above)
Comment (VISHNUDAS-tunerlabs) [reply module/users/helper.js:635]: need to check always passing stateLocationId - with app team

File: controllers/v1/entities.js [hunk 26]
```diff
@@ -152,9 +152,9 @@ module.exports = class Entities extends Abstract {
      */
 
     subEntityList(req) {
-        
+
```
Comment (priyanka-TL) [inline controllers/v1/entities.js:155]: @VISHNUDAS-tunerlabs remove unwanted new line

File: controllers/v1/users.js [hunk 27]
```diff
@@ -556,7 +556,7 @@ module.exports = class Users extends Abstract {
                 else if ( roleWiseTargetedEntities && roleWiseTargetedEntities.length > 1 ) {
                     
                     let targetedEntity = await usersHelper.getHighestTargetedEntity(
-                        roleWiseTargetedEntities
+                        roleWiseTargetedEntities, req.body
```
Comment (priyanka-TL) [inline controllers/v1/users.js:559]: @VISHNUDAS-tunerlabs make req.body as optional if getHighestTargetedEntity used by any other function

File: envVariables.js [hunk 28]
```diff
@@ -129,6 +129,10 @@ let enviromentVariables = {
     "message" : "App Portal base url",
     "optional" : true,
     "default": "https://dev.sunbirded.org"
+  },
+  "SUNBIRD_SERVICE_URL" : {
```
Comment (priyanka-TL) [inline envVariables.js:133]: @VISHNUDAS-tunerlabs USER_SERVICE_URL is already there use that

File: generics/constants/common.js [hunk 29]
```diff
@@ -66,5 +66,10 @@ module.exports = {
   "LINK" : "link",
   "DISCOVERED_BY_ME" : "discoveredByMe",
   "DEFAULT_SUBMISSION_REQUIRED" : 1,
-  "SCHOOL" : "school"
+  "SCHOOL" : "school",
+  "SUBENTITY" :"subEntity_",
+  "SUNBIRD_SERVER_TIMEOUT" : 5000,
```
Comment (priyanka-TL) [inline generics/constants/common.js:71]: @VISHNUDAS-tunerlabs add SUNBIRD_SERVER_TIMEOUT, CACHE_TIME_PERIOD in env only if these values are added in env for other services.

File: generics/services/sunbird.js [hunk 30]
```diff
@@ -0,0 +1,247 @@
+/**
```
Comment (priyanka-TL) [inline generics/services/sunbird.js:1]: @VISHNUDAS-tunerlabs use user service instead of this

File: generics/services/sunbird.js [hunk 31]
```diff
@@ -0,0 +1,247 @@
...
+        if ( searchKey !== "" ) {
+            bodyData["request"]["query"] = searchKey
+        }
+        
+          
+        const url = 
+        sunbirdBaseUrl + constants.endpoints.GET_LOCATION_DATA;
+        const options = {
+            headers : {
+                "Authorization" : process.env.SUNBIRD_SERVICE_AUTHERIZATION,
```
Comment (priyanka-TL) [inline generics/services/sunbird.js:52]: @VISHNUDAS-tunerlabs do we need to pass Authorization for internal call?

File: generics/services/sunbird.js [hunk 32]
```diff
@@ -0,0 +1,247 @@
...
+                    subType: subTypeData,
+                    action: constants.common.GET_METHOD
+                }
+            }
+            
+            const url = 
+            sunbirdBaseUrl + constants.endpoints.GET_FORM_DATA;
+            const options = {
+                headers : {
+                    "Authorization" : process.env.SUNBIRD_SERVICE_AUTHERIZATION,
```
Comment (priyanka-TL) [inline generics/services/sunbird.js:118]: @VISHNUDAS-tunerlabs check Authorization is required

File: generics/services/sunbird.js [hunk 33]
```diff
@@ -0,0 +1,247 @@
...
+                        result["data"] = response.result;
+                        result.success = true;
+                    } else {
+                        result.success = false;
+                    }
+                }
+                return resolve(result);
+            }
+            setTimeout(function () {
+                return reject (result = {
```
Comment (priyanka-TL) [inline generics/services/sunbird.js:147]: @VISHNUDAS-tunerlabs use resolve instead of reject

File: generics/services/sunbird.js [hunk 34]
```diff
@@ -0,0 +1,247 @@
...
+                bodyData["request"]["fields"] = fields;
+            }
+
+        
+            
+            const url = 
+            sunbirdBaseUrl + constants.endpoints.GET_SCHOOL_DATA;
+            const options = {
+                headers : {
+                    "Authorization" : process.env.SUNBIRD_SERVICE_AUTHERIZATION,
```
Comment (priyanka-TL) [inline generics/services/sunbird.js:202]: @VISHNUDAS-tunerlabs do we need Authorization ?

File: generics/services/sunbird.js [hunk 35]
```diff
@@ -0,0 +1,247 @@
...
+                    if( response.responseCode === constants.common.OK) {
+                        result["data"] = response.result;
+                    } else {
+                        result.success = false;
+                    }
+                }
+                return resolve(result);
+            }
+            setTimeout(function () {
+                return reject (result = {
```
Comment (priyanka-TL) [inline generics/services/sunbird.js:232]: @VISHNUDAS-tunerlabs use resolve instead of reject

File: module/entities/helper.js [hunk 36]
```diff
@@ -1,7 +1,11 @@
+
+
 const entityTypesHelper = require(MODULES_BASE_PATH + "/entityTypes/helper");
 const userRolesHelper = require(MODULES_BASE_PATH + "/user-roles/helper");
 // const elasticSearch = require(GENERIC_HELPERS_PATH + "/elastic-search");
-
+const sunbirdService = require(ROOT_PATH + '/generics/services/sunbird');
```
Comment (priyanka-TL) [inline module/entities/helper.js:6]: @VISHNUDAS-tunerlabs use user service

File: module/entities/helper.js [hunk 37]
```diff
@@ -241,59 +246,116 @@ module.exports = class EntitiesHelper {
...
-                    _id: entityId
-                }, projection);
+                let entitiesData = await sunbirdService.learnerLocationSearch(
+                    bodyData,
+                    pageSize,
+                    pageNo,
+                    searchText
+                );
+                
+                if( !entitiesData.success || !entitiesData.data || !entitiesData.data.response.length > 0 ) {
```
Comment (priyanka-TL) [inline module/entities/helper.js:260]: @VISHNUDAS-tunerlabs Is entitiesData.data.response will present always ?

File: module/entities/helper.js [hunk 38]
```diff
@@ -241,59 +246,116 @@ module.exports = class EntitiesHelper {
...
+                let immediateLocation = entitiesData.data.response;
+                
+
+                return resolve({
+                    data : immediateLocation,
+                    count : entitiesData.data.count
+                });
+
+
+                // if( searchText !== "" ){
```
Comment (priyanka-TL) [inline module/entities/helper.js:276]: @VISHNUDAS-tunerlabs remove commented code if its not required

File: module/entities/helper.js [hunk 39]
```diff
@@ -241,59 +246,116 @@ module.exports = class EntitiesHelper {
...
     
-                        immediateEntities = searchImmediateData[0];
-                    }
-                }
-    
-                return resolve(immediateEntities);
+                //         immediateEntities = searchImmediateData[0];
+                //     }
+                // }
+                // console.log("immediateEntities : ",immediateEntities)
```
Comment (priyanka-TL) [inline module/entities/helper.js:357]: @VISHNUDAS-tunerlabs remove console

File: module/entities/helper.js [hunk 40]
```diff
@@ -836,3 +962,54 @@ module.exports = class EntitiesHelper {
   }
 
 }
+
+// /**
```
Comment (priyanka-TL) [inline module/entities/helper.js:966]: @VISHNUDAS-tunerlabs remove the comments

File: module/programs/helper.js [hunk 41]
```diff
@@ -10,6 +10,7 @@
 const entityTypesHelper = require(MODULES_BASE_PATH + "/entityTypes/helper");
 const entitiesHelper = require(MODULES_BASE_PATH + "/entities/helper");
 const userRolesHelper = require(MODULES_BASE_PATH + "/user-roles/helper");
+const sunbirdService = require(ROOT_PATH + '/generics/services/sunbird');
```
Comment (priyanka-TL) [inline module/programs/helper.js:13]: @VISHNUDAS-tunerlabs use user service

File: module/programs/helper.js [hunk 42]
```diff
@@ -721,37 +716,49 @@ module.exports = class ProgramsHelper {
...
-          _id : { $in : entities },
-          entityTypeId : programData[0].scope.entityTypeId
-        },["_id"]);
-          
-        if( !entitiesData.length > 0 ) {
+        let locationIds = [];
+        let bodyData={};
+        
+        entities.forEach(entity=>{
+          if (gen.utils.checkValidUUID(entity)) {
```
Comment (priyanka-TL) [inline module/programs/helper.js:732]: @VISHNUDAS-tunerlabs what if the school code also pass in the req

File: module/programs/helper.js [hunk 43]
```diff
@@ -721,37 +716,49 @@ module.exports = class ProgramsHelper {
...
+            locationIds.push(entity);
+          }
+        });
+        bodyData = {
+          "id" : locationIds
+        }
+        
+        let entitiesData = await sunbirdService.learnerLocationSearch( bodyData );
```
Comment (priyanka-TL) [inline module/programs/helper.js:740]: @VISHNUDAS-tunerlabs remove unwanted new line

File: module/solutions/helper.js [hunk 44]
```diff
@@ -5,6 +5,8 @@
  * Description : Solution related helper functionality.
  */
 
+
+
```
Comment (priyanka-TL) [inline module/solutions/helper.js:9]: @VISHNUDAS-tunerlabs remove unwanted new line

File: module/solutions/helper.js [hunk 45]
```diff
@@ -14,6 +16,7 @@ const userRolesHelper = require(MODULES_BASE_PATH + "/user-roles/helper");
 const surveyService = require(ROOT_PATH + '/generics/services/survey');
 const improvementProjectService = require(ROOT_PATH + '/generics/services/improvement-project');
 const appsPortalBaseUrl = process.env.APP_PORTAL_BASE_URL + "/" ;
+const sunbirdService = require(ROOT_PATH + '/generics/services/sunbird');
```
Comment (priyanka-TL) [inline module/solutions/helper.js:19]: @VISHNUDAS-tunerlabs use user service

File: module/users/helper.js [hunk 46]
```diff
@@ -12,6 +12,9 @@ const userRolesHelper = require(MODULES_BASE_PATH + "/user-roles/helper");
 const entitiesHelper = require(MODULES_BASE_PATH + "/entities/helper");
 const improvementProjectService = require(ROOT_PATH + "/generics/services/improvement-project");
 const userService = require(ROOT_PATH + "/generics/services/users");
+const sunbirdService = require(ROOT_PATH + '/generics/services/sunbird');
```
Comment (priyanka-TL) [inline module/users/helper.js:15]: @VISHNUDAS-tunerlabs use user service

File: controllers/v1/users.js (hunk 27 above)
Comment (VISHNUDAS-tunerlabs) [reply controllers/v1/users.js:559]: getHighestTargetedEntity  used only here

File: module/programs/helper.js [hunk 47]
```diff
@@ -854,50 +861,47 @@ module.exports = class ProgramsHelper {
   static removeEntitiesInScope( programId,entities ) {
     return new Promise(async (resolve, reject) => {
       try {
-
         let programData = 
         await this.programDocuments({ 
           _id : programId,
           scope : { $exists : true },
           isAPrivateProgram : false 
-        },["_id","scope.entityTypeId"]);
```
[the remark this answers is not in the document]
Comment (VISHNUDAS-tunerlabs) [reply module/programs/helper.js:863]: checked
scope : {
        entityType : String,
        entityTypeId : "ObjectId",
        entities : {
          type : Array,
          index : true
        },
        roles : [{
          _id : "ObjectId",
          code : {
            type : String,
            index : true
          }
        }]
      }

File: module/users/helper.js [hunk 48]
```diff
@@ -310,15 +353,15 @@ module.exports = class UsersHelper {
...
       * @method
       * @name entitiesMappingForm
-      * @param {String} stateId - state id.
+      * @param {String} stateCode - state code.
       * @param {String} roleId - role id.
       * @returns {Object} returns a list of entitiesMappingForm.
      */
 
-    static entitiesMappingForm(stateId, roleId) {
+    static entitiesMappingForm(stateCode, roleId, entityKey) {
```
[the remark this answers is not in the document]
Comment (VISHNUDAS-tunerlabs) [reply module/users/helper.js:361]: using only here

**Pull request merged**